### ISSUE TITLE
Can you share the code that includes preprocessing stages like resizing to 712x1072? Because I am getting lower results?

### ISSUE THREAD

**Issue by osivaz61** [guomugong/SEBNet#3]

**Comment by 20011112zx** [comment]
May I ask how the labels ALL of the data are merged and directly stacked during the training process

**Comment by osivaz61** [comment]
import os
import cv2
import glob
import numpy as np
from PIL import Image
from torch.utils.data import Dataset, DataLoader

data_path = 'D:/datasets/IDRID/A. Segmentation/1. Original Images/a. Training Set/'
label_path = 'D:/datasets/IDRID/A. Segmentation/2. All Segmentation Groundtruths/a. Training Set/'

def crop_image_from_gray(image: np.ndarray, maskLabel: np.ndarray, threshold: int = 30):
    if image.ndim == 2:
        mask = image > threshold
        return image[np.ix_(mask.any(1), mask.any(0))]
    elif image.ndim == 3:
        gray_image = cv2.cvtColor(image, cv2.COLOR_RGB2GRAY)
    mask = gray_image > threshold

    check_shape = image[:, :, 0][np.ix_(mask.any(1), mask.any(0))].shape[0]
    if (check_shape == 0):
        return image
    else:
        image1 = image[:, :, 0][np.ix_(mask.any(1), mask.any(0))]
        image2 = image[:, :, 1][np.ix_(mask.any(1), mask.any(0))]
        image3 = image[:, :, 2][np.ix_(mask.any(1), mask.any(0))]
        image = np.stack([image1, image2, image3], axis = -1)

        croppedLabelsList = []
        mask11 = maskLabel[0][:, :, 0][np.ix_(mask.any(1), mask.any(0))]
        mask22 = maskLabel[0][:, :, 1][np.ix_(mask.any(1), mask.any(0))]
        mask33 = maskLabel[0][:, :, 2][np.ix_(mask.any(1), mask.any(0))]
        croppedLabelsList.append(np.stack([mask11, mask22, mask33], axis = -1))

        mask11 = maskLabel[1][:, :, 0][np.ix_(mask.any(1), mask.any(0))]
        mask22 = maskLabel[1][:, :, 1][np.ix_(mask.any(1), mask.any(0))]
        mask33 = maskLabel[1][:, :, 2][np.ix_(mask.any(1), mask.any(0))]
        croppedLabelsList.append(np.stack([mask11, mask22, mask33], axis = -1))

        mask11 = maskLabel[2][:, :, 0][np.ix_(mask.any(1), mask.any(0))]
        mask22 = maskLabel[2][:, :, 1][np.ix_(mask.any(1), mask.any(0))]
        mask33 = maskLabel[2][:, :, 2][np.ix_(mask.any(1), mask.any(0))]
        croppedLabelsList.append(np.stack([mask11, mask22, mask33], axis = -1))

        mask11 = maskLabel[3][:, :, 0][np.ix_(mask.any(1), mask.any(0))]
        mask22 = maskLabel[3][:, :, 1][np.ix_(mask.any(1), mask.any(0))]
        mask33 = maskLabel[3][:, :, 2][np.ix_(mask.any(1), mask.any(0))]
        croppedLabelsList.append(np.stack([mask11, mask22, mask33], axis = -1))

        return image, croppedLabelsList

from PIL import Image
from torchvision.transforms import Resize
class MyIDRiDImageDataset(Dataset):
    def __init__(self, img_dir, labels_img_dir, normalize=False, resize=None):
        self.labels_img_dir = labels_img_dir
        self.img_dir = img_dir
        self.all_images = os.listdir(self.img_dir)
        self.all_labels = os.listdir(self.labels_img_dir + str("1. Microaneurysms/"))
        self.normalize = normalize
        self.resize = resize

    def __len__(self):
        return len(self.all_labels)

    def __getitem__(self, idx):

        img_path = os.path.join(self.img_dir, self.all_images[idx])
        image = cv2.imread(str(img_path))

        tempName = self.all_labels[idx].split(".")[0].split("_")[0] + "_" + self.all_labels[idx].split(".")[0].split("_")[1]

        labelsList = []
        label_ma = cv2.imread(os.path.join(self.labels_img_dir + str("1. Microaneurysms"), self.all_labels[idx]))  ## 54 tane
        labelsList.append(label_ma)

        label_he = cv2.imread(os.path.join(self.labels_img_dir + str("2. Haemorrhages"), tempName + "_HE.tif")) ## 53 tane
        if label_he is None:
            labelsList.append(np.zeros_like(label_ma))
        else:
            labelsList.append(label_he)

        label_ex = cv2.imread(os.path.join(self.labels_img_dir + str("3. Hard Exudates"), tempName + "_EX.tif"))  ## 54 tane
        labelsList.append(label_ex)

        label_se = cv2.imread(os.path.join(self.labels_img_dir + str("4. Soft Exudates"), tempName + "_SE.tif"))  ## 26 tane
        if label_se is None:
            labelsList.append(np.zeros_like(label_ma))
        else:
            labelsList.append(label_se)

        image, labelsList = crop_image_from_gray(image, labelsList)

        image = Resize(size = (768, 1024), interpolation = Image.BICUBIC)(Image.fromarray(np.uint8(image)).convert('RGB'))
        labelsList[0] = Resize(size = (768, 1024), interpolation = Image.NEAREST)(Image.fromarray(np.uint8(labelsList[0])).convert('RGB'))
        labelsList[1] = Resize(size = (768, 1024), interpolation = Image.NEAREST)(Image.fromarray(np.uint8(labelsList[1])).convert('RGB'))
        labelsList[2] = Resize(size = (768, 1024), interpolation = Image.NEAREST)(Image.fromarray(np.uint8(labelsList[2])).convert('RGB'))
        labelsList[3] = Resize(size = (768, 1024), interpolation = Image.NEAREST)(Image.fromarray(np.uint8(labelsList[3])).convert('RGB'))
        
        
        image = np.asarray(image)
        ## image = image.transpose(2, 0, 1)
        labelsList[0] = np.asarray(labelsList[0])
        labelsList[1] = np.asarray(labelsList[1])
        labelsList[2] = np.asarray(labelsList[2])
        labelsList[3] = np.asarray(labelsList[3])

        all_label = np.zeros((768, 1024), dtype = np.uint8)
        all_label[np.where(labelsList[0][:, :, 2] == 255)] = 3
        all_label[np.where(labelsList[1][:, :, 2] == 255)] = 2
        all_label[np.where(labelsList[2][:, :, 2] == 255)] = 1
        all_label[np.where(labelsList[3][:, :, 2] == 255)] = 4

        cv2.imwrite("data/idrid/test768x1024/image/" + tempName + '.tif', image)
        cv2.imwrite("data/idrid/test768x1024/label/" + tempName + '_ALL.tif', np.asarray([all_label, all_label, all_label]).transpose(1, 2, 0))


        return image, labelsList, tempName

class CFG:
    ######################
    # Globals #
    ######################
    seed = 1213
    epochs = 100
    img_size = 512
    main_metric = "epoch_score"
    minimize_metric = False
    num_workers = 4

    ######################
    # Loaders #
    ######################
    loader_params = {
        "train": {
            "batch_size": 1,
            "num_workers": num_workers,
            "shuffle": True
        },
        "valid": {
            "batch_size": 16,
            "num_workers": num_workers,
            "shuffle": False
        },
        "test": {
            "batch_size": 16,
            "num_workers": num_workers,
            "shuffle": False
        }
    }

    ######################
    # Split #
    ######################
    split = "MultilabelStratifiedKFold"
    split_params = {
        "n_splits": 2,
        "shuffle": True,
        "random_state": 1213
    }

    ######################
    # Model #
    ######################
    base_model_name = "tf_efficientnet_b2_ns"
    pooling = "GeM"
    pretrained = True
    num_classes = 1

    ######################
    # Criterion #
    ######################
    loss_name = "BCEFocalLoss"
    loss_params: dict = {}

    ######################
    # Optimizer #
    ######################
    optimizer_name = "Adam"
    base_optimizer = "Adam"
    optimizer_params = {
        "lr": 0.0005
    }
    # For SAM optimizer
    base_optimizer = "Adam"

    ######################
    # Scheduler #
    ######################
    scheduler_name = "CosineAnnealingLR"
    scheduler_params = {
        "T_max": 10
    }

if __name__ == '__main__':

    img_data = MyIDRiDImageDataset(img_dir = 'D:/datasets/IDRID/A. Segmentation/1. Original Images/b. Testing Set/', 
                                   labels_img_dir = 'D:/datasets/IDRID/A. Segmentation/2. All Segmentation Groundtruths/b. Testing Set/',
                                   resize = 512, normalize = True)
    train_dataloader = DataLoader(img_data, batch_size = 1, shuffle = False, num_workers = 4)

    for i, data in enumerate(train_dataloader, 1):
        inputs, labels, fileID = data

**Comment by osivaz61** [comment]
import os
import cv2
import glob
import numpy as np
from PIL import Image
from torch.utils.data import Dataset, DataLoader

data_path = 'D:/datasets/IDRID/A. Segmentation/1. Original Images/a. Training Set/'
label_path = 'D:/datasets/IDRID/A. Segmentation/2. All Segmentation Groundtruths/a. Training Set/'

def crop_image_from_gray(image: np.ndarray, maskLabel: np.ndarray, threshold: int = 30):
    if image.ndim == 2:
        mask = image > threshold
        return image[np.ix_(mask.any(1), mask.any(0))]
    elif image.ndim == 3:
        gray_image = cv2.cvtColor(image, cv2.COLOR_RGB2GRAY)
    mask = gray_image > threshold

    check_shape = image[:, :, 0][np.ix_(mask.any(1), mask.any(0))].shape[0]
    if (check_shape == 0):
        return image
    else:
        image1 = image[:, :, 0][np.ix_(mask.any(1), mask.any(0))]
        image2 = image[:, :, 1][np.ix_(mask.any(1), mask.any(0))]
        image3 = image[:, :, 2][np.ix_(mask.any(1), mask.any(0))]
        image = np.stack([image1, image2, image3], axis = -1)

        croppedLabelsList = []
        mask11 = maskLabel[0][:, :, 0][np.ix_(mask.any(1), mask.any(0))]
        mask22 = maskLabel[0][:, :, 1][np.ix_(mask.any(1), mask.any(0))]
        mask33 = maskLabel[0][:, :, 2][np.ix_(mask.any(1), mask.any(0))]
        croppedLabelsList.append(np.stack([mask11, mask22, mask33], axis = -1))

        mask11 = maskLabel[1][:, :, 0][np.ix_(mask.any(1), mask.any(0))]
        mask22 = maskLabel[1][:, :, 1][np.ix_(mask.any(1), mask.any(0))]
        mask33 = maskLabel[1][:, :, 2][np.ix_(mask.any(1), mask.any(0))]
        croppedLabelsList.append(np.stack([mask11, mask22, mask33], axis = -1))

        mask11 = maskLabel[2][:, :, 0][np.ix_(mask.any(1), mask.any(0))]
        mask22 = maskLabel[2][:, :, 1][np.ix_(mask.any(1), mask.any(0))]
        mask33 = maskLabel[2][:, :, 2][np.ix_(mask.any(1), mask.any(0))]
        croppedLabelsList.append(np.stack([mask11, mask22, mask33], axis = -1))

        mask11 = maskLabel[3][:, :, 0][np.ix_(mask.any(1), mask.any(0))]
        mask22 = maskLabel[3][:, :, 1][np.ix_(mask.any(1), mask.any(0))]
        mask33 = maskLabel[3][:, :, 2][np.ix_(mask.any(1), mask.any(0))]
        croppedLabelsList.append(np.stack([mask11, mask22, mask33], axis = -1))

        return image, croppedLabelsList

from PIL import Image
from torchvision.transforms import Resize
class MyIDRiDImageDataset(Dataset):
    def __init__(self, img_dir, labels_img_dir, normalize=False, resize=None):
        self.labels_img_dir = labels_img_dir
        self.img_dir = img_dir
        self.all_images = os.listdir(self.img_dir)
        self.all_labels = os.listdir(self.labels_img_dir + str("1. Microaneurysms/"))
        self.normalize = normalize
        self.resize = resize

    def __len__(self):
        return len(self.all_labels)

    def __getitem__(self, idx):

        img_path = os.path.join(self.img_dir, self.all_images[idx])
        image = cv2.imread(str(img_path))

        tempName = self.all_labels[idx].split(".")[0].split("_")[0] + "_" + self.all_labels[idx].split(".")[0].split("_")[1]

        labelsList = []
        label_ma = cv2.imread(os.path.join(self.labels_img_dir + str("1. Microaneurysms"), self.all_labels[idx]))  ## 54 tane
        labelsList.append(label_ma)

        label_he = cv2.imread(os.path.join(self.labels_img_dir + str("2. Haemorrhages"), tempName + "_HE.tif")) ## 53 tane
        if label_he is None:
            labelsList.append(np.zeros_like(label_ma))
        else:
            labelsList.append(label_he)

        label_ex = cv2.imread(os.path.join(self.labels_img_dir + str("3. Hard Exudates"), tempName + "_EX.tif"))  ## 54 tane
        labelsList.append(label_ex)

        label_se = cv2.imread(os.path.join(self.labels_img_dir + str("4. Soft Exudates"), tempName + "_SE.tif"))  ## 26 tane
        if label_se is None:
            labelsList.append(np.zeros_like(label_ma))
        else:
            labelsList.append(label_se)

        image, labelsList = crop_image_from_gray(image, labelsList)

        image = Resize(size = (768, 1024), interpolation = Image.BICUBIC)(Image.fromarray(np.uint8(image)).convert('RGB'))
        labelsList[0] = Resize(size = (768, 1024), interpolation = Image.NEAREST)(Image.fromarray(np.uint8(labelsList[0])).convert('RGB'))
        labelsList[1] = Resize(size = (768, 1024), interpolation = Image.NEAREST)(Image.fromarray(np.uint8(labelsList[1])).convert('RGB'))
        labelsList[2] = Resize(size = (768, 1024), interpolation = Image.NEAREST)(Image.fromarray(np.uint8(labelsList[2])).convert('RGB'))
        labelsList[3] = Resize(size = (768, 1024), interpolation = Image.NEAREST)(Image.fromarray(np.uint8(labelsList[3])).convert('RGB'))
        
        
        image = np.asarray(image)
        ## image = image.transpose(2, 0, 1)
        labelsList[0] = np.asarray(labelsList[0])
        labelsList[1] = np.asarray(labelsList[1])
        labelsList[2] = np.asarray(labelsList[2])
        labelsList[3] = np.asarray(labelsList[3])

        all_label = np.zeros((768, 1024), dtype = np.uint8)
        all_label[np.where(labelsList[0][:, :, 2] == 255)] = 3
        all_label[np.where(labelsList[1][:, :, 2] == 255)] = 2
        all_label[np.where(labelsList[2][:, :, 2] == 255)] = 1
        all_label[np.where(labelsList[3][:, :, 2] == 255)] = 4

        cv2.imwrite("data/idrid/test768x1024/image/" + tempName + '.tif', image)
        cv2.imwrite("data/idrid/test768x1024/label/" + tempName + '_ALL.tif', np.asarray([all_label, all_label, all_label]).transpose(1, 2, 0))


        return image, labelsList, tempName

class CFG:
    ######################
    # Globals #
    ######################
    seed = 1213
    epochs = 100
    img_size = 512
    main_metric = "epoch_score"
    minimize_metric = False
    num_workers = 4

    ######################
    # Loaders #
    ######################
    loader_params = {
        "train": {
            "batch_size": 1,
            "num_workers": num_workers,
            "shuffle": True
        },
        "valid": {
            "batch_size": 16,
            "num_workers": num_workers,
            "shuffle": False
        },
        "test": {
            "batch_size": 16,
            "num_workers": num_workers,
            "shuffle": False
        }
    }

    ######################
    # Split #
    ######################
    split = "MultilabelStratifiedKFold"
    split_params = {
        "n_splits": 2,
        "shuffle": True,
        "random_state": 1213
    }

    ######################
    # Model #
    ######################
    base_model_name = "tf_efficientnet_b2_ns"
    pooling = "GeM"
    pretrained = True
    num_classes = 1

    ######################
    # Criterion #
    ######################
    loss_name = "BCEFocalLoss"
    loss_params: dict = {}

    ######################
    # Optimizer #
    ######################
    optimizer_name = "Adam"
    base_optimizer = "Adam"
    optimizer_params = {
        "lr": 0.0005
    }
    # For SAM optimizer
    base_optimizer = "Adam"

    ######################
    # Scheduler #
    ######################
    scheduler_name = "CosineAnnealingLR"
    scheduler_params = {
        "T_max": 10
    }

if __name__ == '__main__':

    img_data = MyIDRiDImageDataset(img_dir = 'D:/datasets/IDRID/A. Segmentation/1. Original Images/b. Testing Set/', 
                                   labels_img_dir = 'D:/datasets/IDRID/A. Segmentation/2. All Segmentation Groundtruths/b. Testing Set/',
                                   resize = 512, normalize = True)
    train_dataloader = DataLoader(img_data, batch_size = 1, shuffle = False, num_workers = 4)

    for i, data in enumerate(train_dataloader, 1):
        inputs, labels, fileID = data

**Comment by osivaz61** [comment]
all_label = np.zeros((768, 1024), dtype = np.uint8)
        all_label[np.where(labelsList[0][:, :, 2] == 255)] = 3
        all_label[np.where(labelsList[1][:, :, 2] == 255)] = 2
        all_label[np.where(labelsList[2][:, :, 2] == 255)] = 1  
        all_label[np.where(labelsList[3][:, :, 2] == 255)] = 4
